routing/datasource.go

```diff
@@ -353,48 +353,73 @@ func mapPredicates(cps []PredicateSpec) map[string]PredicateSpec {
 }
 
 // processes a set of route definitions for the routing table
-func processRouteDefs(o Options, fr filters.Registry, defs []*eskip.Route) []*Route {
+func processRouteDefs(o Options, fr filters.Registry, defs []*eskip.Route) (routes []*Route, invalidDefs []*eskip.Route) {
 	cpm := mapPredicates(o.Predicates)
-
-	var routes []*Route
 	for _, def := range defs {
 		route, err := processRouteDef(cpm, fr, def)
 		if err == nil {
 			routes = append(routes, route)
 		} else {
+			invalidDefs = append(invalidDefs, def)
 			o.Log.Error(err)
 		}
 	}
+	return
+}
 
-	return routes
+type routeTable struct {
+	m             *matcher
+	validRoutes   []*Route
+	invalidRoutes []*Route
+	invalidDefs   []*eskip.Route
+	created       time.Time
 }
 
 // receives the next version of the routing table on the output channel,
 // when an update is received on one of the data clients.
-func receiveRouteMatcher(o Options, out chan<- *matcher, quit <-chan struct{}) {
+func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}) {
 	updates := receiveRouteDefs(o, quit)
 	var (
-		mout         *matcher
-		outRelay     chan<- *matcher
+		rt           *routeTable
+		outRelay     chan<- *routeTable
 		updatesRelay <-chan []*eskip.Route
 	)
-
 	updatesRelay = updates
 	for {
 		select {
 		case defs := <-updatesRelay:
 			o.Log.Info("route settings received")
-			routes := processRouteDefs(o, o.FilterRegistry, defs)
+			routes, invalidDefs := processRouteDefs(o, o.FilterRegistry, defs)
 			m, errs := newMatcher(routes, o.MatchingOptions)
+
+			invalidRouteIds := make(map[string]struct{})
+			invalidRoutes := []*Route{}
+			validRoutes := []*Route{}
+
 			for _, err := range errs {
 				o.Log.Error(err)
+				invalidRouteIds[err.Id] = struct{}{}
```

Comment:
However we still need to extract valid and invalid routes, so I'm not sure returning invalid Route will make the need for this map to go away.